File: server/datasets/java_source-code/Multiple-G3/renamed_40_dsu.java
public class DSU_4 {
    public static void main(String[] args) {
        DSU_4 dsu = new DSU_4(10);
        dsu.u_4(1, 2);
    }
    int[] p_4;
    int[] r_4;
    public DSU_4(int tot_4) {
        p_4 = new int[tot_4];
        r_4 = new int[tot_4];
        for (int idx = 0; idx < tot_4; idx++) {
            p_4[idx] = idx;
            r_4[idx] = 0;
        }
    }
    public int f_4(int x_4) {
        if (p_4[x_4] != x_4) {
            p_4[x_4] = f_4(p_4[x_4]);
        }
        return p_4[x_4];
    }
    public void u_4(int a_4, int b_4) {
        int ra_4 = f_4(a_4);
        int rb_4 = f_4(b_4);
        if (ra_4 != rb_4) {
            if (r_4[ra_4] < r_4[rb_4]) {
                p_4[ra_4] = rb_4;
            } else if (r_4[ra_4] > r_4[rb_4]) {
                p_4[rb_4] = ra_4;
            } else {
                p_4[rb_4] = ra_4;
                r_4[ra_4]++;
            }
        }
    }
}

File: server/datasets/python_source-code/Multiple-G3/renamed_40_dsu.py
class DSU_4:
    def __init__(self, n_4):
        self.p_4 = list(range(n_4))
        self.rk_4 = [0] * n_4

    def find_4(self, x_4):
        if self.p_4[x_4] != x_4:
            self.p_4[x_4] = self.find_4(self.p_4[x_4])
        return self.p_4[x_4]

    def union_4(self, x_4_a, x_4_b):
        ra_4 = self.find_4(x_4_a)
        rb_4 = self.find_4(x_4_b)
        if ra_4 != rb_4:
            if self.rk_4[ra_4] < self.rk_4[rb_4]:
                ra_4, rb_4 = rb_4, ra_4
            self.p_4[rb_4] = ra_4
            if self.rk_4[ra_4] == self.rk_4[rb_4]:
                self.rk_4[ra_4] += 1
            return True
        return False
